example/prometheus/main.go

```diff
@@ -45,7 +45,7 @@ func initMeter() *push.Controller {
 	if err != nil {
 		log.Panicf("failed to initialize metric stdout exporter %v", err)
 	}
-	batcher := defaultkeys.New(selector, sdkmetric.NewDefaultLabelEncoder(), false)
+	batcher := defaultkeys.New(selector, sdkmetric.NewDefaultLabelEncoder(), true)
```

Comment:
I like what you've done here as far as moving the code forward and using the OTel aggregators!

I'm glad to see `stateful` being used here, as this seemed like the reason it exists. As I read the Prometheus-Go client for summary and histogram, though, both types reset with each collection. We want stateful in the sense that we should remember all metrics _and_ maintain sums for counters / last values for gauges, but measure state resets on each collection.

If measure state didn't reset on each collection, the distribution would grow to summarize a longer and longer period of time as the process ages, to the point where new data will have very little impact on the output.

It doesn't "feel right" to me, to use the push controller and save a reference to the checkpoint set, but I think it can work if they are synchronized.

In the PR description you talk about "lighter" queries, but I think that calling `Collect()` more often than needed is actually going to cost more CPU in total, i.e., you're doing more work that necessary. It makes sense to block the query to run collection as this is the lowest cost approach--it has fewer calls into the Batcher.

Also, I think we want the collection to slow down when the system is overloaded. If we can't collect data fast enough, the collection interval should increase.

When we add Observer metric instruments, they will naturally get called once per interval as a result of calling Collect() once per query (https://github.com/open-telemetry/oteps/pull/72)--this is another consideration coming ahead.

I think it is appropriate to implement an integrated Batcher and Controller that do this logic needed here without convoluted synchronization or letting the exporter mutate the checkpointSet (which this calls for, it seems).  One approach would be to create a new type of stateful Batcher that embeds a stateful DefaultKeys batcher but provides an alternate `FinishedCollection` method that would reset Aggregator state for measures only. There's still some need for synchronization between the Batcher/Exporter/Controller, which is why I suggest integrating all three.

Reply:
(My suggestion about FinishedCollection doesn't really work, but providing some kind of configurable hook to mutate state at the end of a collection probably does.) It's an unusual requirement, as we want to remember all metrics but only reset some of their state with each collection.

Reply:
Thanks for taking a look at this @jmacd, I'll try to answer your quentions here.

`histogram` are just a collection of counters for each bucket, a counter for the total events observed and the sum of all observed values. These types should not reset after each collection.

`summary` on the other hand is a bit more trickier, I've overlooked this. The prometheus' clients does not completely reset them. Summaries are calculated using a rolling window of quantiles with configured number of windows and the max "age" of each window. The default values for these are 5 windows and 10 minutes max age, resulting in 5 2-minute quantile windows.

This feels very particular stateful batchers. I don't know if this type of rolling window for summaries should be kept on the SDK or on the prometheus exporter, thought on this?

Thse are some cool articles from mechanics of `Histogram` and `Summary`:
1. [How does a Prometheus Histogram work?](https://www.robustperception.io/how-does-a-prometheus-histogram-work)
2. [How does a Prometheus Summary work?](https://www.robustperception.io/how-does-a-prometheus-summary-work)

Reply:
Yeah -- I had probably misunderstood the prometheus-golang histogram code, e.g., 
https://github.com/prometheus/client_golang/blob/e7776d2c54305c1b62fdb113b5a7e9b944c5c27e/prometheus/histogram.go#L338

(As I study it however, it's a bit fishy -- this looks like it will lose cold counts in a race.
```
		atomic.AddUint64(&hotCounts.buckets[i], atomic.LoadUint64(&coldCounts.buckets[i]))
		atomic.StoreUint64(&coldCounts.buckets[i], 0)
```)

Reply:
I don't see how this could lose cold counts. From what I could understand, only the `Write` method changes the could counts and it is guarded by a mutex. 

Btw, that algorithm could be a good solution to a more correct lock free minmaxsumcount aggregator and maybe a lock free ddsketch aggregator as well.

Reply:
OK. It's just fishy to see an atomic read followed by an atomic store (if there's a mutex protecting it, why use atomics?). Nevermind.

Also: ACK re: the algorithm.

Reply:
That sounds like a useful place to add more documentation if it causes programmer WTFs.

Reply:
Yes, a comment would be useful about why a stateful Batcher is used. We should encapsulate this in a simpler installation helper (#345). 

I'd be happy for you to file an issue about potentially using the atomic `countAndHotIdx` algorithm used in Prometheus -- maybe the code could be packaged in a way that it's used by the MinMaxSumCount aggregator, Histogram buckets, possibly others.

Also let's add a TODO about how the summary export supported here is not resetting the state at Checkpoint (as far as I know), but it's OK with me to get this code merged and fi that in a future PR as we iterate.

Reply:
I'm not sure the `countAndHotIdx` technique will work for DDSketch, which still needs a lock to protect its internals. It would give lock-freedom for the observer vs. the collector, but not for observer vs. observer.

Reply:
(Anyway, it's a cool technique for Histogram buckets and MinMaxSumCount.)

Reply:
I've added some docs to the reason we need a stateful batcher here. This should be moved to the simplified constructor once that is done.

And will file these issues soon.